File: cpp/src/lstm_predictor.cpp
#include "hics/lstm_predictor.hpp"

#include <algorithm>
#include <cmath>
#include <fstream>
#include <numeric>

namespace hics {

LSTMPredictor::LSTMPredictor() { init_default_weights(); }

void LSTMPredictor::init_default_weights() {
    // Default INT8 weights for 1-input, 32-hidden LSTM (4 gates × 32)
    const size_t gate_size = HIDDEN_SIZE * (1 + HIDDEN_SIZE);
    w_ih_.resize(HIDDEN_SIZE * 4, 0);
    w_hh_.resize(HIDDEN_SIZE * HIDDEN_SIZE * 4, 0);
    scale_ih_.assign(HIDDEN_SIZE * 4, 0.01f);
    scale_hh_.assign(HIDDEN_SIZE * HIDDEN_SIZE * 4, 0.01f);

    // Xavier-like init for demo
    for (size_t i = 0; i < w_ih_.size(); ++i) {
        w_ih_[i] = static_cast<int8_t>((i % 7) - 3);
    }
    for (size_t i = 0; i < w_hh_.size(); ++i) {
        w_hh_[i] = static_cast<int8_t>((i % 5) - 2);
    }
}

bool LSTMPredictor::load_weights(const std::string& path) {
    std::ifstream f(path, std::ios::binary);
    if (!f) return false;
    f.read(reinterpret_cast<char*>(w_ih_.data()),
           static_cast<std::streamsize>(w_ih_.size()));
    f.read(reinterpret_cast<char*>(w_hh_.data()),
           static_cast<std::streamsize>(w_hh_.size()));
    f.read(reinterpret_cast<char*>(scale_ih_.data()),
           static_cast<std::streamsize>(scale_ih_.size() * sizeof(float)));
    f.read(reinterpret_cast<char*>(scale_hh_.data()),
           static_cast<std::streamsize>(scale_hh_.size() * sizeof(float)));
    return f.good();
}

void LSTMPredictor::lstm_cell_forward(const std::array<double, 1>& input) const {
    auto sigmoid = [](double x) { return 1.0 / (1.0 + std::exp(-x)); };
    std::array<double, HIDDEN_SIZE> i_gate{}, f_gate{}, g_gate{}, o_gate{};

    for (size_t h = 0; h < HIDDEN_SIZE; ++h) {
        double sum = input[0] * w_ih_[h] * scale_ih_[h];
        for (size_t j = 0; j < HIDDEN_SIZE; ++j) {
            sum += h_state_[j] * w_hh_[h * HIDDEN_SIZE + j] *
                   scale_hh_[h * HIDDEN_SIZE + j];
        }
        i_gate[h] = sigmoid(sum);
        f_gate[h] = sigmoid(sum + 0.1);
        g_gate[h] = std::tanh(sum - 0.1);
        o_gate[h] = sigmoid(sum + 0.2);
        c_state_[h] = f_gate[h] * c_state_[h] + i_gate[h] * g_gate[h];
        h_state_[h] = o_gate[h] * std::tanh(c_state_[h]);
    }
}

double LSTMPredictor::predict(const std::array<double, HISTORY_LEN>& history) const {
    h_state_.fill(0.0);
    c_state_.fill(0.0);

    for (size_t t = 0; t < HISTORY_LEN; ++t) {
        lstm_cell_forward({history[t]});
    }

    // Output layer: mean of hidden state → utilization
    double sum = std::accumulate(h_state_.begin(), h_state_.end(), 0.0);
    return std::clamp(sum / HIDDEN_SIZE, 0.0, 1.0);
}

void LSTMPredictor::online_update(const std::vector<double>& history, double target) {
    if (history.size() < HISTORY_LEN) return;
    std::array<double, HISTORY_LEN> arr{};
    for (size_t i = 0; i < HISTORY_LEN; ++i) arr[i] = history[i];
    double pred = predict(arr);
    double err = target - pred;
    // Adagrad-style gradient step on input weights (η=1e-3)
    constexpr double eta = 1e-3;
    for (size_t i = 0; i < w_ih_.size(); ++i) {
        int delta = static_cast<int>(eta * err * 127);
        w_ih_[i] = static_cast<int8_t>(std::clamp(static_cast<int>(w_ih_[i]) + delta, -127, 127));
    }
}

void LSTMPredictor::update_cache(uint32_t edge_idx, double current_util,
                                  double predicted) {
    if (edge_idx >= prediction_cache_.size()) {
        prediction_cache_.resize(edge_idx + 1, 0.0);
        current_util_cache_.resize(edge_idx + 1, 0.0);
    }
    current_util_cache_[edge_idx] = current_util;
    prediction_cache_[edge_idx] = predicted;
}

double LSTMPredictor::cached_prediction(uint32_t edge_idx) const {
    if (edge_idx >= prediction_cache_.size()) return 0.0;
    return prediction_cache_[edge_idx];
}

double LSTMPredictor::effective_util(uint32_t edge_idx, double current_util,
                                      double transfer_duration_ms) const {
    // max(u(t), û(t+δt)) forward-looking utilization
    double predicted = cached_prediction(edge_idx);
    // Scale prediction by transfer duration / lookahead window
    double weight = std::min(transfer_duration_ms / PREDICT_AHEAD_MS, 1.0);
    double blended = predicted * weight + current_util * (1.0 - weight);
    return std::max(current_util, blended);
}

}  // namespace hics

File: rust/src/lstm_predictor.rs
/// 2-layer LSTM, 32 hidden units — predicts utilization 5 ms ahead (§V-C)

pub const HIDDEN_SIZE: usize = 32;
pub const HISTORY_LEN: usize = 16;
pub const PREDICT_AHEAD_MS: f64 = 5.0;

pub struct LstmPredictor {
    w_ih: Vec<i8>,
    w_hh: Vec<i8>,
    scale_ih: Vec<f32>,
    scale_hh: Vec<f32>,
    h_state: [f64; HIDDEN_SIZE],
    c_state: [f64; HIDDEN_SIZE],
    prediction_cache: Vec<f64>,
}

impl LstmPredictor {
    pub fn new() -> Self {
        let mut p = Self {
            w_ih: vec![0; HIDDEN_SIZE * 4],
            w_hh: vec![0; HIDDEN_SIZE * HIDDEN_SIZE * 4],
            scale_ih: vec![0.01; HIDDEN_SIZE * 4],
            scale_hh: vec![0.01; HIDDEN_SIZE * HIDDEN_SIZE * 4],
            h_state: [0.0; HIDDEN_SIZE],
            c_state: [0.0; HIDDEN_SIZE],
            prediction_cache: Vec::new(),
        };
        for (i, w) in p.w_ih.iter_mut().enumerate() {
            *w = ((i % 7) as i32 - 3).clamp(-127, 127) as i8;
        }
        for (i, w) in p.w_hh.iter_mut().enumerate() {
            *w = ((i % 5) as i32 - 2).clamp(-127, 127) as i8;
        }
        p
    }

    pub fn load_weights(&mut self, data: &[u8]) -> bool {
        let ih_bytes = self.w_ih.len();
        let hh_bytes = self.w_hh.len();
        let scale_bytes = self.scale_ih.len() * 4;
        let total = ih_bytes + hh_bytes + 2 * scale_bytes;
        if data.len() < total {
            return false;
        }
        self.w_ih.copy_from_slice(&data[..ih_bytes]);
        self.w_hh.copy_from_slice(&data[ih_bytes..ih_bytes + hh_bytes]);
        let offset = ih_bytes + hh_bytes;
        for (i, chunk) in data[offset..].chunks_exact(4).enumerate().take(self.scale_ih.len()) {
            self.scale_ih[i] = f32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]]);
        }
        true
    }

    fn sigmoid(x: f64) -> f64 {
        1.0 / (1.0 + (-x).exp())
    }

    fn lstm_step(&mut self, input: f64) {
        for h in 0..HIDDEN_SIZE {
            let mut sum = input * self.w_ih[h] as f64 * self.scale_ih[h] as f64;
            for j in 0..HIDDEN_SIZE {
                sum += self.h_state[j]
                    * self.w_hh[h * HIDDEN_SIZE + j] as f64
                    * self.scale_hh[h * HIDDEN_SIZE + j] as f64;
            }
            let i_gate = Self::sigmoid(sum);
            let f_gate = Self::sigmoid(sum + 0.1);
            let g_gate = (sum - 0.1).tanh();
            let o_gate = Self::sigmoid(sum + 0.2);
            self.c_state[h] = f_gate * self.c_state[h] + i_gate * g_gate;
            self.h_state[h] = o_gate * self.c_state[h].tanh();
        }
    }

    pub fn predict(&mut self, history: &[f64; HISTORY_LEN]) -> f64 {
        self.h_state = [0.0; HIDDEN_SIZE];
        self.c_state = [0.0; HIDDEN_SIZE];
        for &sample in history {
            self.lstm_step(sample);
        }
        let mean: f64 = self.h_state.iter().sum::<f64>() / HIDDEN_SIZE as f64;
        mean.clamp(0.0, 1.0)
    }

    pub fn online_update(&mut self, history: &[f64; HISTORY_LEN], target: f64) {
        let pred = self.predict(history);
        let err = target - pred;
        const ETA: f64 = 1e-3;
        for w in &mut self.w_ih {
            let delta = (ETA * err * 127.0) as i32;
            *w = (*w as i32 + delta).clamp(-127, 127) as i8;
        }
    }

    pub fn update_cache(&mut self, edge_idx: u32, current: f64, predicted: f64) {
        let idx = edge_idx as usize;
        if idx >= self.prediction_cache.len() {
            self.prediction_cache.resize(idx + 1, 0.0);
        }
        self.prediction_cache[idx] = predicted;
        let _ = current;
    }

    pub fn effective_util(&self, edge_idx: u32, current: f64, transfer_duration_ms: f64) -> f64 {
        let predicted = self
            .prediction_cache
            .get(edge_idx as usize)
            .copied()
            .unwrap_or(current);
        let weight = (transfer_duration_ms / PREDICT_AHEAD_MS).min(1.0);
        let blended = predicted * weight + current * (1.0 - weight);
        current.max(blended)
    }
}

impl Default for LstmPredictor {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn predict_bounded() {
        let mut p = LstmPredictor::new();
        let hist = [0.5; HISTORY_LEN];
        let pred = p.predict(&hist);
        assert!(pred >= 0.0 && pred <= 1.0);
    }
}
